### PR TITLE
Row-group-level partitioning for Parquet

### PR DESCRIPTION
**Pull request by calebwin**
This PR introduces a `row_group_cols` parameter for `cudf.to_parquet` that groups data by the given columns are writes separate columns to separate row groups (row groups are groups of rows within a Parquet file). This is similar to `partition_cols` except instead of separate groups being written to separate files, separate groups are written to separate row groups within a file.

You should use `row_group_cols` when you want to partition data on a column but there are too many groups (or combinations of groups if you are partitioning on a secondary column) that would result in too many small files.

What's left:
- [x] Initial implementation in cudf
- [x] Integration in dask_cudf
- [x] Testing, tests
- [ ] Benchmarking

Notes to reviewers:
- There are changes in cudf-cpp, cudf-python, dask-cudf-python.
- If `dask_cudf.to_parquet` accepts `partition_on` and `row_group_cols`, should `row_group_cols` be renamed to `row_group_on` for API consistency?
- Is the way I am launching the `populate_chunk_hash_maps` and `get_dictionary_indices` kernels correct?
- There is an extra index column written when specifying `row_group_cols` - is this a bug we should fix before merging? It seems to be related to [1] and [2].

[1] https://issues.apache.org/jira/browse/ARROW-9136
[2] https://github.com/pandas-dev/pandas/issues/34790.

### CODE REVIEW

**Comment by jakirkham** [comment]
cc @rjzamora

File: cpp/src/io/parquet/chunk_dict.cu [hunk 1]
```diff
@@ -95,16 +95,18 @@ struct map_find_fn {
...
   auto t       = threadIdx.x;
 
-  auto start_row =
-    block_x *
-    max_page_fragment_size;  // This is fragment size. all chunks are multiple of these many rows.
-  size_type end_row = min(start_row + max_page_fragment_size, num_rows);
+  auto start_row = 0;
+  for (auto i = 0; i < block_x; i++) {
+    start_row += fragments[0][i].num_rows;
+  }
```
Comment (bdice) [inline cpp/src/io/parquet/chunk_dict.cu:108]: Perhaps these offsets should be pre-computed with a scan and then passed into the kernel? I'm not sure how many row groups we expect. The difference between 10 and 1M would indicate whether this should be a host or device computation.

If we shouldn't use a scan and pass in the precomputed offsets, then this could use `std::accumulate`. Might look something like this snippet (untested):

```suggestion
  auto row_counter = thrust::transform_iterator(fragments[0], [] __device__(auto const& page){ return page.num_rows; });
  auto start_row = std::accumulate(row_counter[0], row_counter[block_x], 0);
```

(Note: `page` might not be the right name for the function argument, I am just guessing from `device_2dspan<PageFragment>`)

**Comment by devavret** [comment]
> * Is the way I am launching the `populate_chunk_hash_maps` and `get_dictionary_indices` kernels correct?

Yes. in fact that exactly how I did it in this partitioning PR #9810 https://github.com/rapidsai/cudf/blob/200d1b0f65d89822d1f1a1016ad3ec73ca2b378f/cpp/src/io/parquet/chunk_dict.cu#L303

File: cpp/src/io/parquet/chunk_dict.cu [hunk 2]
```diff
@@ -245,14 +247,18 @@ __global__ void __launch_bounds__(block_size, 1)
...
   auto col_idx = blockIdx.y;
   auto block_x = blockIdx.x;
   auto t       = threadIdx.x;
 
-  size_type start_row = block_x * max_page_fragment_size;
-  size_type end_row   = min(start_row + max_page_fragment_size, num_rows);
+  auto start_row = 0;
+  for (auto i = 0; i < block_x; i++) {
+    start_row += fragments[0][i].num_rows;
+  }
```
Comment (bdice) [inline cpp/src/io/parquet/chunk_dict.cu:260]: Same as previous comment.

File: cpp/src/io/parquet/writer_impl.cu [hunk 3]
```diff
@@ -20,6 +20,7 @@
  */
 
 #include <io/statistics/column_statistics.cuh>
+#include "io/parquet/parquet_gpu.hpp"
```
Comment (bdice) [inline cpp/src/io/parquet/writer_impl.cu:23]: This should probably use `<>` braces.
```suggestion
#include <io/parquet/parquet_gpu.hpp>
```

File: cpp/src/io/parquet/writer_impl.cu [hunk 4]
```diff
@@ -32,7 +33,9 @@
 #include <cudf/null_mask.hpp>
 #include <cudf/strings/strings_column_view.hpp>
 #include <cudf/table/table_device_view.cuh>
+#include <cudf/utilities/span.hpp>
 
+#include <iostream>
```
Comment (bdice) [inline cpp/src/io/parquet/writer_impl.cu:38]: Looks like this was left in from some print debugging? If it is needed, it should go with the other section of stdlib headers like `#include <algorithm>` below here, rather than with the rmm includes.

File: cpp/src/io/parquet/writer_impl.cu [hunk 5]
```diff
@@ -1152,45 +1158,93 @@ void writer::impl::write(table_view const& table)
...
+    num_fragments = 0;
+    for (std::size_t i = 0; i < row_group_sizes.size(); i++) {
+      num_fragments += (row_group_sizes[i] + max_page_fragment_size - 1) / max_page_fragment_size;
+    }
+  }
   cudf::detail::hostdevice_2dvector<gpu::PageFragment> fragments(
     num_columns, num_fragments, stream);
 
+  if (row_group_sizes_specified) {
+    // auto fragments_span = host_2dspan<gpu::PageFragment>{fragments};
```
Comment (bdice) [inline cpp/src/io/parquet/writer_impl.cu:1173]: Why is this commented?

File: cpp/src/io/parquet/writer_impl.cu [hunk 6]
```diff
@@ -1152,45 +1158,93 @@ void writer::impl::write(table_view const& table)
   // compression/decompression performance).
   using cudf::io::parquet::gpu::max_page_fragment_size;
 
-  size_type const num_fragments = (num_rows + max_page_fragment_size - 1) / max_page_fragment_size;
+  bool row_group_sizes_specified = !row_group_sizes.empty();
+  size_type num_fragments        = (num_rows + max_page_fragment_size - 1) / max_page_fragment_size;
```
Comment (bdice) [inline cpp/src/io/parquet/writer_impl.cu:1162]: Is this the same math as `div_rounding_up_safe`? That utility function might be applicable in a few places here.

File: cpp/src/io/parquet/writer_impl.cu [hunk 7]
```diff
@@ -1152,45 +1158,93 @@ void writer::impl::write(table_view const& table)
...
+    for (std::size_t i = 0; i < row_group_sizes.size(); i++) {
+      num_fragments += (row_group_sizes[i] + max_page_fragment_size - 1) / max_page_fragment_size;
+    }
+  }
   cudf::detail::hostdevice_2dvector<gpu::PageFragment> fragments(
     num_columns, num_fragments, stream);
 
+  if (row_group_sizes_specified) {
+    // auto fragments_span = host_2dspan<gpu::PageFragment>{fragments};
+    size_type i = 0;
```
Comment (bdice) [inline cpp/src/io/parquet/writer_impl.cu:1174]: Can we use more meaningful names for the indices here? It doesn't appear that this function is using a traditional 3d indexing scheme like `arr[i][j][k]`, and I found the uses of `i`, `j`, `k` somewhat hard to follow without knowing what they represent.

File: python/cudf/cudf/io/parquet.py [hunk 8]
```diff
@@ -73,6 +74,9 @@ def write_to_dataset(
     partition_cols : list,
         Column names by which to partition the dataset
         Columns are partitioned in the order they are given
+    row_group_cols : list,
```
Comment (bdice) [inline python/cudf/cudf/io/parquet.py:77]: 1. The optional values in this docstring should specify their default values and optionality.

See https://pandas.pydata.org/pandas-docs/version/1.1.5/reference/api/pandas.DataFrame.to_parquet.html
> partition_cols : list, optional, default None

2. Options like `root_path` and `**kwargs` with no default values should not have a trailing comma after the type name.

3. Types labeled as `string` should be labeled as `str` instead.

File: python/cudf/cudf/io/parquet.py [hunk 9]
```diff
@@ -120,12 +124,18 @@ def write_to_dataset(
...
                         )
                     )
                 else:
-                    write_df.to_parquet(fil, index=preserve_index, **kwargs)
+                    write_df.to_parquet(
+                        fil,
+                        index=preserve_index,
+                        row_group_cols=row_group_cols,
+                        **kwargs,
+                    )
```
Comment (bdice) [inline python/cudf/cudf/io/parquet.py:138]: I think this section of code can be written to only call `to_parquet` once. Something roughly like this, which updates `**kwargs` and optionally keeps the result:

```python
if return_metadata:
    kwargs["metadata_file_path"] = fs.sep.join([subdir, filename])

metadata_result = write_df.to_parquet(
    fil,
    index=preserve_index,
    row_group_cols=row_group_cols,
    **kwargs,
)

if return_metadata:
    metadata.append(metadata_result)
```

File: python/cudf/cudf/tests/test_parquet.py [hunk 10]
```diff
@@ -428,6 +428,48 @@ def num_row_groups(rows, group_size):
...
+
+    fname = tmpdir.join("row_group_sizes.parquet")
+    df.to_parquet(fname, compression="snappy", row_group_sizes=row_group_sizes)
+
+    num_rows, row_groups, col_names = cudf.io.read_parquet_metadata(fname)
+
+    assert num_rows == len(df.index)
+    assert row_groups == len(row_group_sizes)
+    for a, b in zip(col_names, df.columns):
+        assert a == b
```
Comment (bdice) [inline python/cudf/cudf/tests/test_parquet.py:449]: Is this equivalent?
```suggestion
    assert col_names == df.columns
```

File: python/cudf/cudf/tests/test_parquet.py [hunk 11]
```diff
@@ -428,6 +428,48 @@ def num_row_groups(rows, group_size):
...
+        columns=pd.Index(list("abc")),
+    )
+
+    fname = tmpdir.join("row_group_sizes.parquet")
+    df.to_parquet(fname, compression="snappy", row_group_sizes=row_group_sizes)
+
+    num_rows, row_groups, col_names = cudf.io.read_parquet_metadata(fname)
+
+    assert num_rows == len(df.index)
+    assert row_groups == len(row_group_sizes)
```
Comment (bdice) [inline python/cudf/cudf/tests/test_parquet.py:447]: This line verifies the _number_ of row groups, but I don't think this test is checking the number of rows in each row group. That seems important to test here.

File: python/cudf/cudf/utils/ioutils.py [hunk 12]
```diff
@@ -215,6 +215,10 @@
     will be written to different directories, but all files will
     have this name.  If nothing is specified, a random uuid4 hex string
     will be used for each file.
+row_group_cols : list, optional, default None
+    Column names by which to partition the dataset across row groups in the
+    resulting Parquet file
+    Columns are partitioned in the order they are given
```
Comment (bdice) [inline python/cudf/cudf/utils/ioutils.py:221]: Sentences in docstrings should end in a period. Line breaks should be avoided except to soft-wrap.
```suggestion
    Column names by which to partition the dataset across row groups in the
    resulting Parquet file. Columns are partitioned in the order they are
    given.
```

File: python/dask_cudf/dask_cudf/io/tests/test_parquet.py [hunk 13]
```diff
@@ -309,6 +310,27 @@ def test_roundtrip_from_dask_partitioned(tmpdir, parts, daskcudf, metadata):
...
+    df["month"] = [1, 2, 3, 3, 3, 2]
+    df["day"] = [1, 1, 1, 2, 2, 1]
+    df["data"] = [0, 0, 0, 0, 0, 0]
+
+    ddf = dask_cudf.from_cudf(cudf.from_pandas(df), npartitions=2)
+    ddf.to_parquet(
+        tmpdir, partition_on=partition_on, row_group_cols=row_group_cols
+    )
+    ddf_read = dask_cudf.read_parquet(tmpdir)
+    assert_eq(len(ddf), len(ddf_read))
```
Comment (bdice) [inline python/dask_cudf/dask_cudf/io/tests/test_parquet.py:331]: `assert_eq` is intended for more complicated assertions about dataframes being equivalent. Comparing lengths should be done with a plain `assert`. However, it is probaby a good idea to make sure the data frame written/read is equivalent to the source dataframe in memory:
```suggestion
    assert len(ddf) == len(ddf_read)
    assert_eq(ddf, ddf_read)
```

**Review by bdice — CHANGES_REQUESTED** [review]
Nice feature @calebwin! I have some comments below.

File: cpp/src/io/parquet/chunk_dict.cu (hunk 1 above)
Comment (devavret) [reply cpp/src/io/parquet/chunk_dict.cu:108]: Do you mean like this:https://github.com/rapidsai/cudf/blob/f44a50b0bf723941a577bf6d71c3e00ba825d231/cpp/src/io/parquet/chunk_dict.cu#L107

Comment (bdice) [reply cpp/src/io/parquet/chunk_dict.cu:108]: @devavret Looks about right! I'm just trying to avoid a loop on each thread when we could use a single-pass scan ahead of time. I see you've worked on this in #9810. That logic should be used here. Does #9810 need to be merged first?

Comment (devavret) [reply cpp/src/io/parquet/chunk_dict.cu:108]: > Does #9810 need to be merged first?

Actually, that's what I was wondering just now. #9810 is close to completion and if it is merged first, then there will be many merge issues with this PR. I'm fine with merging #9810 later or taking over this one if it remains unmerged due to merge issues.

cc @quasiben @vuule

File: cpp/src/io/parquet/page_enc.cu [hunk 14]
```diff
@@ -130,7 +130,11 @@ __global__ void __launch_bounds__(block_size)
   if (!t) {
     // frag.num_rows = fragment_size except for the last page fragment which can be smaller.
     // num_rows is fixed but fragment size could be larger if the data is strings or nested.
-    s->frag.num_rows           = min(fragment_size, max_num_rows - min(start_row, max_num_rows));
+    if (fragment_size != -1) {
+      s->frag.num_rows = min(fragment_size, max_num_rows - min(start_row, max_num_rows));
+    } else {
+      s->frag.num_rows = frag[blockIdx.x][blockIdx.y].num_rows;
+    }
```
Comment (hyperbolic2346) [inline cpp/src/io/parquet/page_enc.cu:137]:
```suggestion
    s->frag.num_rows = fragment_size != -1 ? min(fragment_size, max_num_rows - min(start_row, max_num_rows)) : frag[blockIdx.x][blockIdx.y].num_rows;
```

File: cpp/src/io/parquet/page_enc.cu [hunk 15]
```diff
@@ -1938,13 +1942,13 @@ dremel_data get_dremel_data(column_view h_col,
...
  * @param[in] col_desc Column description array [column_id]
- * @param[in] num_fragments Number of fragments per column
+ * @param[in] num_fragments Number of fragments per column, -1 if fragment sizes already specified
  * @param[in] num_columns Number of columns
  * @param[in] stream CUDA stream to use, default 0
  */
 void InitPageFragments(device_2dspan<PageFragment> frag,
                        device_span<parquet_column_device_view const> col_desc,
-                       uint32_t fragment_size,
+                       int32_t fragment_size,
```
Comment (hyperbolic2346) [inline cpp/src/io/parquet/page_enc.cu:1951]: Are we concerned about the loss of resolution here?

File: cpp/src/io/parquet/writer_impl.cu [hunk 16]
```diff
@@ -1152,45 +1158,93 @@ void writer::impl::write(table_view const& table)
   // compression/decompression performance).
   using cudf::io::parquet::gpu::max_page_fragment_size;
 
-  size_type const num_fragments = (num_rows + max_page_fragment_size - 1) / max_page_fragment_size;
+  bool row_group_sizes_specified = !row_group_sizes.empty();
+  size_type num_fragments        = (num_rows + max_page_fragment_size - 1) / max_page_fragment_size;
+  if (row_group_sizes_specified) {
+    num_fragments = 0;
+    for (std::size_t i = 0; i < row_group_sizes.size(); i++) {
+      num_fragments += (row_group_sizes[i] + max_page_fragment_size - 1) / max_page_fragment_size;
```
Comment (hyperbolic2346) [inline cpp/src/io/parquet/writer_impl.cu:1166]: This should be a `std::accumulate` as Bradley showed above.

**Review by hyperbolic2346 — CHANGES_REQUESTED** [review]
Thanks for working on this. Some comments. Interested to see this progress.

File: cpp/src/io/parquet/chunk_dict.cu (hunk 1 above)
Comment (vuule) [reply cpp/src/io/parquet/chunk_dict.cu:108]: @devavret do you have an ETA for addressing current comments on #9810?

File: cpp/src/io/parquet/page_enc.cu (hunk 15 above)
Comment (bdice) [reply cpp/src/io/parquet/page_enc.cu:1951]: I asked a similar question on another PR and @vuule said it's fine. https://github.com/rapidsai/cudf/pull/9677#discussion_r748668781

File: cpp/src/io/parquet/chunk_dict.cu (hunk 1 above)
Comment (vuule) [reply cpp/src/io/parquet/chunk_dict.cu:108]: My 2cts:
Let's aim to merge this first so Caleb has a chance to get the PR as close as possible to the finish line as possible. If 9810 already addresses some comments here, maybe those pieces can be applied to this PR (also reduces merge conflicts).

I'm not sure what's the best approach, inclined to leave the decision up to @devavret and @calebwin .

Comment (calebwin) [reply cpp/src/io/parquet/chunk_dict.cu:108]: Thanks @bdice @hyperbolic2346 @devavret @vuule for reviews and comments. I just ran into a subtle CUDA bug in this PR when I was in the middle of writing a benchmark. I looked through #9810 and it looks like there are changes in the CUDA code that may handle edge cases I didn't consider here.

So I'm going to go ahead and try to merge #9810 into this and make appropriate changes. I will see if that resolves the issue I came across when benchmarking. I will then try to address other reviews here.

Should I convert this PR to draft in the meanwhile?

**Comment by vyasr** [comment]
I'm going to close this since it's fairly out of date and it's not clear if we still want this as is. Feel free to reopen if work on this restarts.
